drivers/sensor/tdk/icm42688/icm42688_decoder.c

```diff
@@ -234,81 +234,33 @@ static inline q31_t icm42688_read_temperature_from_packet(const uint8_t *pkt)
 static int icm42688_read_imu_from_packet(const uint8_t *pkt, bool is_accel, int fs,
 					 uint8_t axis_offset, q31_t *out)
 {
-	int32_t value;
-	int64_t scale = 0;
-	int32_t max = BIT(15);
+	uint32_t unsigned_value;
+	int32_t signed_value;
+	bool is_hires = FIELD_GET(FIFO_HEADER_ACCEL, pkt[0]) == 1;
 	int offset = 1 + (axis_offset * 2);
 
-	if (is_accel) {
-		switch (fs) {
-		case ICM42688_DT_ACCEL_FS_2:
-			scale = INT64_C(2) * BIT(31 - 5) * 9.80665;
-			break;
-		case ICM42688_DT_ACCEL_FS_4:
-			scale = INT64_C(4) * BIT(31 - 6) * 9.80665;
-			break;
-		case ICM42688_DT_ACCEL_FS_8:
-			scale = INT64_C(8) * BIT(31 - 7) * 9.80665;
-			break;
-		case ICM42688_DT_ACCEL_FS_16:
-			scale = INT64_C(16) * BIT(31 - 8) * 9.80665;
-			break;
-		}
-	} else {
-		switch (fs) {
-		case ICM42688_DT_GYRO_FS_2000:
-			scale = 164;
-			break;
-		case ICM42688_DT_GYRO_FS_1000:
-			scale = 328;
-			break;
-		case ICM42688_DT_GYRO_FS_500:
-			scale = 655;
-			break;
-		case ICM42688_DT_GYRO_FS_250:
-			scale = 1310;
-			break;
-		case ICM42688_DT_GYRO_FS_125:
-			scale = 2620;
-			break;
-		case ICM42688_DT_GYRO_FS_62_5:
-			scale = 5243;
-			break;
-		case ICM42688_DT_GYRO_FS_31_25:
-			scale = 10486;
-			break;
-		case ICM42688_DT_GYRO_FS_15_625:
-			scale = 20972;
-			break;
-		}
-	}
+	const uint32_t scale[2][2] = {
+		/* low-res,	hi-res */
+		{35744,		8936}, /* gyro */
```

Comment:
Having the scales precalculated in a table is great, but at least a formula for how these are calculated should be in a comment here

Reply:
Really, I think I think a macro could be useful here for computing a scaling factor

E.g.

SENSOR_Q31_SCALER(sensitivity, shift) is half the battle, e.g. getting the math for most sensor conversions looking like the simple integer multiplier math that it should be, could be done simpler I think in most cases. Better still I love const table here, though sadly C doesn't let us do const math expressions like the following even if the compiler will do it at compile time.

```c
const sensitivity = SENSITIVITIES[drv_cfg->fs];
const shift = SHIFTS[drv_cfg->fs];
const int32_t scaler = (1/sensitivity)*(pow(2,31)/pow(2,shift));
q31_t scaled = sample*scaler;
```

There's also occasionally the unit conversion involved, e.g. converting G to m/s^2 would involve another multiplier of 9.81 m/s^2. But clearly it was wrong here and needed to be rectified.